### PR TITLE
Add queue to send email notifications

### PR DESCRIPTION
**Pull request by caiseL**

### CODE REVIEW

File: server/src/main/resources/evolutions/default/1.sql [hunk 1]
```diff
@@ -42,3 +42,21 @@ CREATE TABLE user_tokens (
...
 CREATE INDEX user_tokens_user_id_index ON user_tokens USING BTREE (user_id);
 
+--contains the status for notifications
+CREATE TABLE user_notifications (
+    user_notification_id UUID NOT NULL,
+    user_id UUID NOT NULL,
+    notification_type TEXT NOT NULL,
+    subject TEXT NOT NULL,
+    message TEXT NOT NULL,
+    status TEXT NOT NULL, -- pending/success/failed,
```
Comment (edgmAlarcon) [inline server/src/main/resources/evolutions/default/1.sql:52]: please create a type for status. for example
```CREATE TYPE NOTIFICATION_STATUS AS ENUM (
    'PENDING',
    'FAILED',
    'SUCCESS'
);```

Comment (AlexITC) [reply server/src/main/resources/evolutions/default/1.sql:52]: I asked Cristopher to not use postgres enums for now, we can add those once we are confident that no more changes will be required.

File: server/src/main/resources/evolutions/default/1.sql [hunk 2]
```diff
@@ -42,3 +42,21 @@ CREATE TABLE user_tokens (
...
+    status_details TEXT NULL, -- if failed, what was the reason
+    error_count INT DEFAULT 0,
+    execute_at TIMESTAMPTZ NOT NULL,
+    created_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
+    updated_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
+    CONSTRAINT user_notifications_user_notification_id_pk PRIMARY KEY (user_notification_id),
+    CONSTRAINT user_notifications_user_id_fk FOREIGN KEY (user_id) REFERENCES users(user_id)
+);
+
+CREATE INDEX user_notifications_user_id_index ON user_notifications USING BTREE (user_id);
```
Comment (AlexITC) [inline server/src/main/resources/evolutions/default/1.sql:62]: We'll need an index for `execute_at`, we'll need filter the tasks to execute by this field

File: server/src/main/scala/net/wiringbits/actions/CreateUserAction.scala [hunk 3]
```diff
@@ -41,16 +39,13 @@ class CreateUserAction @Inject() (
...
-      // then, send the verification email
       emailParameter = s"${createUser.id}_$token"
-      emailRequest = EmailMessage.registration(
+      emailMessage = EmailMessage.registration(
         name = createUser.name,
         url = webAppConfig.host,
         emailParameter = emailParameter
       )
-      _ = emailApi.sendEmail(EmailRequest(request.email, emailRequest))
+      _ <- usersRepository.create(createUser, emailMessage)
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/actions/CreateUserAction.scala:48]: We need to keep sending the email in this scenario, otherwise, the token will be logged in plain text in the database

File: server/src/main/scala/net/wiringbits/actions/ForgotPasswordAction.scala [hunk 4]
```diff
@@ -53,8 +51,7 @@ class ForgotPasswordAction @Inject() (
     ValidateVerifiedUser(user)
     val emailMessage = EmailMessage.forgotPassword(user.name, webAppConfig.host, s"${user.id}_$token")
     for {
-      _ <- userTokensRepository.create(createToken)
-      _ = emailApi.sendEmail(EmailRequest(user.email, emailMessage))
+      _ <- userTokensRepository.create(createToken, emailMessage)
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/actions/ForgotPasswordAction.scala:54]: Same

File: server/src/main/scala/net/wiringbits/tasks/NotificationsTask.scala [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    getPendingNotifications: GetPendingNotificationsAction,
+    sendNotificationAction: SendNotificationAction
+)(implicit
+    ec: ExecutionContext,
+    actorSystem: ActorSystem
+) {
+  val logger = LoggerFactory.getLogger(this.getClass)
+
+  logger.info("Starting the notifications task")
+  actorSystem.scheduler.scheduleAtFixedRate(
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/tasks/NotificationsTask.scala:22]: Let's use `scheduleOnce` instead, when it finishes we just schedule it once again, this way, we won't have two concurrent tasks running concurrently which will likely break the workflow, [example](https://github.com/X9Developers/block-explorer/blob/develop/server/app/com/xsn/explorer/tasks/PollerSynchronizerTask.scala#L33)

Also, in order to write test, we will have to create a `run` method that returns a handler to stop the task but we can check that later.

File: server/src/main/scala/net/wiringbits/tasks/NotificationsTask.scala [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    notificationsConfig.delayedInit,
+    notificationsConfig.interval
+  ) { () =>
+    run()
+  }
+
+  def run(): Unit = {
+    getPendingNotifications.apply.foreach { notifications =>
+      logger.info(s"There's ${notifications.size} pending notifications")
+      notifications.foreach(notification => sendNotificationAction(notification))
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/tasks/NotificationsTask.scala:32]: `sendNotificationAction` returns a future but the value is being discarded, we must log the outcome when it is an error

File: server/src/main/scala/net/wiringbits/repositories/models/NotificationType.scala [hunk 7]
```diff
@@ -0,0 +1,16 @@
+package net.wiringbits.repositories.models
+
+import enumeratum.EnumEntry.Uppercase
+import enumeratum.{Enum, EnumEntry}
+
+sealed trait NotificationType extends EnumEntry with Uppercase
+
+object NotificationType extends Enum[NotificationType] {
+  final case object EmailRegistration extends NotificationType
+  final case object VerifyAccount extends NotificationType
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/repositories/models/NotificationType.scala:10]: `VerifyAccount`/`ResetPassword` won't get into the table, we need to take only the notifications that do not involve confidential data

File: server/src/main/scala/net/wiringbits/repositories/daos/UserNotificationsDAO.scala [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+import java.util.UUID
+
+object UserNotificationsDAO {
+
+  import anorm._
+
+  def create(request: UserNotification.Create)(implicit conn: Connection): Unit = {
+    val _ = SQL"""
+      INSERT INTO user_notifications
+        (user_notification_id, user_id, notification_type, subject, message, status, execute_at)
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/repositories/daos/UserNotificationsDAO.scala:16]: Let's include `created_at`/`updated_at`, I'm trying to avoid using timestamps from SQL because it complicates writing tests

File: server/src/main/scala/net/wiringbits/repositories/daos/UserNotificationsDAO.scala [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+      """
+      .execute()
+  }
+
+  def getPendingNotifications()(implicit conn: Connection): List[UserNotification] = {
+    SQL"""
+      SELECT user_notification_id, user_id, notification_type, subject, message, status, status_details, error_count, execute_at, created_at, updated_at
+      FROM user_notifications
+      WHERE status != ${NotificationStatus.Success.toString}
+        AND execute_at <= NOW()
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/repositories/daos/UserNotificationsDAO.scala:35]: Let's prefer `clock.instant` over `NOW()`.

We could define the max errors where we won't try anymore, possibly 10 is good enough.

File: server/src/main/scala/net/wiringbits/repositories/daos/UserNotificationsDAO.scala [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+  def setStatusToFailed(notificationId: UUID, executeAt: Instant, failReason: String)(implicit
+      conn: Connection
+  ): Unit = {
+    val _ = SQL"""
+      UPDATE user_notifications SET 
+        status = ${NotificationStatus.Failed.toString}::TEXT,
+        status_details = $failReason,
+        error_count = error_count + 1,
+        execute_at = $executeAt::TIMESTAMPTZ,
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/repositories/daos/UserNotificationsDAO.scala:48]: `updated_at` needs to be updated.

File: server/src/main/scala/net/wiringbits/repositories/daos/UserNotificationsDAO.scala [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+      .execute()
+  }
+
+  def setStatusToSuccess(notificationId: UUID)(implicit
+      conn: Connection
+  ): Unit = {
+    val _ = SQL"""
+      UPDATE user_notifications SET 
+        status = ${NotificationStatus.Success.toString}::TEXT
+      WHERE user_notification_id = ${notificationId.toString}::UUID
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/repositories/daos/UserNotificationsDAO.scala:60]: Same.

File: server/src/main/scala/net/wiringbits/repositories/UserTokensRepository.scala [hunk 12]
```diff
@@ -1,23 +1,37 @@
...
     database: Database
 )(implicit
-    ec: DatabaseExecutionContext
+    ec: DatabaseExecutionContext,
+    clock: Clock
 ) {
 
-  def create(request: UserToken.Create): Future[Unit] = Future {
-    database.withConnection { implicit conn =>
+  def create(request: UserToken.Create, emailMessage: EmailMessage): Future[Unit] = Future {
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/repositories/UserTokensRepository.scala:21]: This shouldn't go here, tokens do not go in plain text to the database.

File: server/src/main/scala/net/wiringbits/config/NotificationsConfig.scala [hunk 13]
```diff
@@ -0,0 +1,20 @@
+package net.wiringbits.config
+
+import play.api.Configuration
+
+import scala.concurrent.duration.FiniteDuration
+
+case class NotificationsConfig(delayedInit: FiniteDuration, interval: FiniteDuration, delayIfFailure: FiniteDuration) {
+  override def toString: String = {
+    s"NotificationsConfig(delayedInit = $delayedInit, interval = $interval, delayIfFailure = $delayIfFailure)"
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/config/NotificationsConfig.scala:9]: `delayIfFailure` is not needed, let's use an exponential backoff strategy starting from 1 minute: 1m, 2m, 4m, 8m, 16m, etc.

File: server/src/main/scala/net/wiringbits/apis/EmailApiAWSImpl.scala [hunk 14]
```diff
@@ -42,8 +43,12 @@ class EmailApiAWSImpl @Inject() (
         .destination(destination)
         .message(message)
         .build()
-      client.sendEmail(request)
-      logger.info(s"Email sent, to: ${emailRequest.destination}, subject = ${emailRequest.message.subject}")
+      for {
+        response <- client.sendEmail(request).toScala
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/apis/EmailApiAWSImpl.scala:47]: This is ideal to be in a different PR to get it merged soon.

File: server/src/main/scala/net/wiringbits/actions/VerifyUserEmailAction.scala [hunk 15]
```diff
@@ -35,7 +32,7 @@ class VerifyUserEmailAction @Inject() (
     _ = ValidateUserToken(userToken)
 
     // then, the user is marked as verified
-    _ <- usersRepository.verify(userId = userId, tokenId = userToken.id)
-    _ = emailApi.sendEmail(EmailRequest(user.email, EmailMessage.confirm(user.name)))
+    emailMessage = EmailMessage.confirm(user.name)
+    _ <- usersRepository.verify(userId = userId, tokenId = userToken.id, emailMessage)
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/actions/VerifyUserEmailAction.scala:36]: This is a good case for the background job

File: server/src/main/scala/net/wiringbits/actions/UpdatePasswordAction.scala [hunk 16]
```diff
@@ -13,18 +11,16 @@ import javax.inject.Inject
...
 
   def apply(userId: UUID, request: UpdatePassword.Request): Future[Unit] = {
     for {
       maybe <- usersRepository.find(userId)
       user = ValidatePasswordMatches(maybe, request.oldPassword)
       hashedPassword = BCrypt.hashpw(request.newPassword.string, BCrypt.gensalt())
-      _ <- usersRepository.updatePassword(userId, hashedPassword)
       emailMessage = EmailMessage.updatePassword(user.name)
-      _ = emailApi.sendEmail(EmailRequest(user.email, emailMessage))
+      _ <- usersRepository.updatePassword(userId, hashedPassword, emailMessage)
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/actions/UpdatePasswordAction.scala:23]: This is a good use case for the background job

File: server/src/main/scala/net/wiringbits/actions/GetPendingNotificationsAction.scala [hunk 17]
```diff
@@ -0,0 +1,13 @@
+package net.wiringbits.actions
+
+import net.wiringbits.repositories.UserNotificationsRepository
+import net.wiringbits.repositories.models.UserNotification
+
+import javax.inject.Inject
+import scala.concurrent.Future
+
+class GetPendingNotificationsAction @Inject() (userNotificationsRepository: UserNotificationsRepository) {
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/actions/GetPendingNotificationsAction.scala:9]: I like the idea to keep a single action per class, please use the `net.wiringbits.actions.internal` package instead to differentiate what relates to the end user APIs

File: server/src/main/scala/net/wiringbits/actions/ResetPasswordAction.scala [hunk 18]
```diff
@@ -38,11 +35,8 @@ class ResetPasswordAction @Inject() (
       // We trigger the reset password flow
       userMaybe <- usersRepository.find(userId)
       user = userMaybe.getOrElse(throw new RuntimeException(s"User with id $userId wasn't found"))
-      _ <- usersRepository.resetPassword(userId, hashedPassword)
-
-      // Sending a confirmation email
       emailMessage = EmailMessage.resetPassword(user.name)
-      _ = emailApi.sendEmail(EmailRequest(user.email, emailMessage))
+      _ <- usersRepository.resetPassword(userId, hashedPassword, emailMessage)
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/actions/ResetPasswordAction.scala:39]: This is a good use case for the  background job

File: server/src/main/scala/net/wiringbits/actions/SendNotificationAction.scala [hunk 19]
```diff
@@ -0,0 +1,43 @@
+package net.wiringbits.actions
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/actions/SendNotificationAction.scala:1]: Same about the `internal` package

File: server/src/main/scala/net/wiringbits/actions/SendNotificationAction.scala [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+    userNotificationsRepository: UserNotificationsRepository,
+    notificationsConfig: NotificationsConfig
+)(implicit
+    ec: ExecutionContext,
+    clock: Clock
+) {
+
+  def apply(notification: UserNotification): Future[Unit] = {
+    try {
+      sendEmail(notification)
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/actions/SendNotificationAction.scala:27]: This won't work because the method returns a future, this is what you need

```
sendEmail(...).onComplete {
  case Failure(ex) => userNotificationsRepository.setStatusToFailed(...)
  case Success(_) =>()
}
```

**Review by AlexITC — CHANGES_REQUESTED** [review]
This is a good start, still, there is stuff to be polished.

Thanks.

File: server/src/main/scala/net/wiringbits/tasks/NotificationsTask.scala [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+
+  logger.info("Starting the notifications task")
+  actorSystem.scheduler.scheduleOnce(
+    notificationsConfig.interval
+  ) {
+    run()
+  }
+
+  def run(): Unit = {
+    logger.info("Looking for notifications")
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/tasks/NotificationsTask.scala:30]: This is good for debugging but noisy when everything works, either remove it or use `logger.trace`

File: server/src/main/scala/net/wiringbits/tasks/NotificationsTask.scala [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+    getPendingNotifications()
+      .onComplete {
+        case Failure(exception) => logger.error("Failed to get notifications", exception)
+        case Success(notifications) =>
+          logger.info(s"There's ${notifications.size} pending notifications")
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/tasks/NotificationsTask.scala:35]: Same way, this can be very noisy when there are no notifications, I'd do this:

```
if (notifications.isEmpty) logger.trace(...)
else logger.info(...)
```

File: server/src/main/scala/net/wiringbits/tasks/NotificationsTask.scala [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+          notifications.foreach { notification =>
+            {
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/tasks/NotificationsTask.scala:37]: This `{}` is unnecessary

**Review by AlexITC — COMMENTED** [review]
It is almost there, nice job.

File: server/src/main/scala/net/wiringbits/actions/internal/SendNotificationAction.scala [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+    usersRepository: UsersRepository,
+    userNotificationsRepository: UserNotificationsRepository
+)(implicit
+    ec: ExecutionContext,
+    clock: Clock
+) {
+
+  def apply(notification: UserNotification): Future[Unit] = {
+    sendEmail(notification).recoverWith { case NonFatal(ex) =>
+      val minutesUntilExecute = DelayGenerator.createDelay(notification.errorCount)
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/actions/internal/SendNotificationAction.scala:26]: Let's add a warning log when the operation failed

File: server/src/main/scala/net/wiringbits/repositories/daos/UserNotificationsDAO.scala [hunk 25]
```diff
@@ -0,0 +1,71 @@
...
+  def getPendingNotifications(
+      allowedErrors: Int = 10
+  )(implicit conn: Connection, clock: Clock): List[UserNotification] = {
+    SQL"""
+      SELECT user_notification_id, user_id, notification_type, subject, message, status, status_details, error_count, execute_at, created_at, updated_at
+      FROM user_notifications
+      WHERE status != ${NotificationStatus.Success.toString}
+        AND execute_at <= ${clock.instant()}
+        AND error_count < $allowedErrors 
+      ORDER BY created_at, user_notification_id
```
Comment (AlexITC) [inline server/src/main/scala/net/wiringbits/repositories/daos/UserNotificationsDAO.scala:41]: hmm, I'd sort by `execute_at`

File: server/src/test/scala/utils/DelayGeneratorSpec.scala [hunk 26]
```diff
@@ -0,0 +1,20 @@
+package utils
```
Comment (AlexITC) [inline server/src/test/scala/utils/DelayGeneratorSpec.scala:1]: `net.wiringbits.util`

**Review by AlexITC — APPROVED** [review]
